src/code/z_message_PAL.c

```diff
@@ -1,66 +1,3306 @@
+    static s16 sAnalogStickHeld = false;
+
+    MessageContext* msgCtx = &globalCtx->msgCtx;
+    Input* curInput = &globalCtx->state.input[0];
+
+    if (curInput->rel.stick_y > 29 && !sAnalogStickHeld) {
+        sAnalogStickHeld = true;
+        msgCtx->choiceIndex--;
+        if (msgCtx->choiceIndex > 128) {
+            msgCtx->choiceIndex = 0;
+        } else {
+            Audio_PlaySoundGeneral(NA_SE_SY_CURSOR, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+        }
+    } else if (curInput->rel.stick_y < -29 && !sAnalogStickHeld) {
+        sAnalogStickHeld = true;
+        msgCtx->choiceIndex++;
+        if (msgCtx->choiceIndex > numChoices) {
+            msgCtx->choiceIndex = numChoices;
+        } else {
+            Audio_PlaySoundGeneral(NA_SE_SY_CURSOR, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+        }
+    } else if (ABS(curInput->rel.stick_y) < 30) {
+        sAnalogStickHeld = false;
+    }
+    msgCtx->textPosX = R_TEXT_CHOICE_XPOS;
+    msgCtx->textPosY =
+        (numChoices == 1) ? R_TEXT_CHOICE_YPOS(msgCtx->choiceIndex + 1) : R_TEXT_CHOICE_YPOS(msgCtx->choiceIndex);
+}
+
+void Message_DrawTextChar(GlobalContext* globalCtx, void* textureImage, Gfx** p) {
+    MessageContext* msgCtx = &globalCtx->msgCtx;
+    Gfx* gfx = *p;
+    s16 x = msgCtx->textPosX;
+    s16 y = msgCtx->textPosY;
+
+    gDPPipeSync(gfx++);
+
+    sCharTexSize = (R_TEXT_CHAR_SCALE / 100.0f) * 16.0f;
+    sCharTexScale = 1024.0f / (R_TEXT_CHAR_SCALE / 100.0f);
+
+    gDPLoadTextureBlock_4b(gfx++, textureImage, G_IM_FMT_I, FONT_CHAR_TEX_WIDTH, FONT_CHAR_TEX_HEIGHT, 0,
+                           G_TX_NOMIRROR | G_TX_CLAMP, G_TX_NOMIRROR | G_TX_CLAMP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD,
+                           G_TX_NOLOD);
+
+    // Draw drop shadow
+    if (msgCtx->textBoxType != BOX_NONE_NO_SHADOW) {
+        gDPSetPrimColor(gfx++, 0, 0, 0, 0, 0, msgCtx->textColorAlpha);
+        gSPTextureRectangle(gfx++, (x + R_TEXT_DROP_SHADOW_OFFSET) << 2, (y + R_TEXT_DROP_SHADOW_OFFSET) << 2,
+                            (x + R_TEXT_DROP_SHADOW_OFFSET + sCharTexSize) << 2,
+                            (y + R_TEXT_DROP_SHADOW_OFFSET + sCharTexSize) << 2, G_TX_RENDERTILE, 0, 0, sCharTexScale,
+                            sCharTexScale);
+    }
+
+    gDPPipeSync(gfx++);
+    gDPSetPrimColor(gfx++, 0, 0, msgCtx->textColorR, msgCtx->textColorG, msgCtx->textColorB, msgCtx->textColorAlpha);
+    gSPTextureRectangle(gfx++, x << 2, y << 2, (x + sCharTexSize) << 2, (y + sCharTexSize) << 2, G_TX_RENDERTILE, 0, 0,
+                        sCharTexScale, sCharTexScale);
+    *p = gfx;
+}
+
+// resizes textboxes when opening them
+void Message_GrowTextbox(MessageContext* msgCtx) {
+    static f32 sWidthCoefficients[] = {
+        1.2f, 1.5f, 1.8f, 2.0f, 2.1f, 2.2f, 2.1f, 2.0f,
+    };
+    static f32 sHeightCoefficients[] = {
+        0.6f, 0.75f, 0.9f, 1.0f, 1.05f, 1.1f, 1.05f, 1.0f,
+    };
+
+    f32 width =
```

Comment:
wont comment on any more, but in general you dont need to separate static vars from stack vars